File: microservice-architecture/load-balancing/python/test/test_balancer.py
import os
import sys
import unittest
from collections import Counter

sys.path.append(os.getcwd())

from src.balancer import (  # noqa: E402
    Backend,
    LeastConnectionsBalancer,
    RoundRobinBalancer,
    WeightedRoundRobinBalancer,
)


class LoadBalancingTest(unittest.TestCase):
    def test_round_robin(self) -> None:
        balancer = RoundRobinBalancer(
            [Backend("node-a"), Backend("node-b"), Backend("node-c")]
        )

        self.assertEqual("node-a", balancer.next().backend_id)
        self.assertEqual("node-b", balancer.next().backend_id)
        self.assertEqual("node-c", balancer.next().backend_id)

    def test_weighted_round_robin(self) -> None:
        balancer = WeightedRoundRobinBalancer(
            [Backend("node-a", weight=2), Backend("node-b", weight=1)]
        )
        counts = Counter(balancer.next().backend_id for _ in range(6))
        self.assertEqual(4, counts["node-a"])
        self.assertEqual(2, counts["node-b"])

    def test_least_connections(self) -> None:
        balancer = LeastConnectionsBalancer(
            [
                Backend("node-a", active_connections=2),
                Backend("node-b", active_connections=0),
                Backend("node-c", active_connections=1),
            ]
        )
        first = balancer.acquire()
        self.assertEqual("node-b", first.backend_id)
        balancer.release("node-b")


if __name__ == "__main__":
    unittest.main()

File: microservice-architecture/load-balancing/js/test/test_balancer.js
import {
  LeastConnectionsBalancer,
  RoundRobinBalancer,
  WeightedRoundRobinBalancer
} from '../src/balancer.js'

function assertEqual(expected, actual, message) {
  if (expected !== actual) {
    throw new Error(`${message} expected=${expected} actual=${actual}`)
  }
}

const roundRobin = new RoundRobinBalancer([
  { backendId: 'node-a' },
  { backendId: 'node-b' },
  { backendId: 'node-c' }
])
assertEqual('node-a', roundRobin.next().backendId, 'round robin first')
assertEqual('node-b', roundRobin.next().backendId, 'round robin second')
assertEqual('node-c', roundRobin.next().backendId, 'round robin third')

const weighted = new WeightedRoundRobinBalancer([
  { backendId: 'node-a', weight: 2 },
  { backendId: 'node-b', weight: 1 }
])
const counts = {}
for (let i = 0; i < 6; i++) {
  const backendId = weighted.next().backendId
  counts[backendId] = (counts[backendId] || 0) + 1
}
assertEqual(4, counts['node-a'], 'node-a weighted count')
assertEqual(2, counts['node-b'], 'node-b weighted count')

const least = new LeastConnectionsBalancer([
  { backendId: 'node-a', activeConnections: 2 },
  { backendId: 'node-b', activeConnections: 0 },
  { backendId: 'node-c', activeConnections: 1 }
])
assertEqual('node-b', least.acquire().backendId, 'least busy backend')
least.release('node-b')

console.log('load-balancing(js) tests passed')
